xbmc/utils/StringUtils.cpp

```diff
@@ -1123,6 +1123,171 @@ int64_t StringUtils::AlphaNumericCompare(const wchar_t* left, const wchar_t* rig
   return 0; // files are the same
 }
 
+/*
+ Convert the UTF8 character to which z points into a 31-bit Unicode point.
+ Return how many bytes (0 to 3) of UTF8 data encode the character.
+ This only works right if z points to a well-formed UTF8 string.
+  Byte-0    Byte-1    Byte-2    Byte-3     Value
+  0xxxxxxx                                 00000000 00000000 0xxxxxxx
+  110yyyyy  10xxxxxx                       00000000 00000yyy yyxxxxxx
+  1110zzzz  10yyyyyy  10xxxxxx             00000000 zzzzyyyy yyxxxxxx
+  11110uuu  10uuzzzz  10yyyyyy  10xxxxxx   000uuuuu zzzzyyyy yyxxxxxx
+*/
+static uint32_t Utf8ToUnicode(const unsigned char* z, int nKey, unsigned char& bytes)
+{
+  // Lookup table used decode the first byte of a multi-byte UTF8 character
+  // clang format off
+  static const unsigned char utf8Trans1[] = {
+    0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+    0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f,
+    0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17,
+    0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f,
+    0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+    0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f,
+    0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
+    0x00, 0x01, 0x02, 0x03, 0x00, 0x01, 0x00, 0x00,
+  };
+  // clang format on
+
+  uint32_t c;
+  int index = 0;
+  c = z[index];
+  index++;
+  if (c >= 0xc0)
+  {
+    c = utf8Trans1[c - 0xc0];
+    while (index < nKey && (z[index] & 0xc0) == 0x80)
+    {
+      c = (c << 6) + (0x3f & z[index]);
+      index++;
+    }
+    if (c < 0x80 || (c & 0xFFFFF800) == 0xD800 || (c & 0xFFFFFFFE) == 0xFFFE)
+      c = 0xFFFD;
+  }
+  bytes = static_cast<unsigned char>(index - 1);
+  return c;
+}
+
+// SQLite collating function, see sqlite3_create_collation
+// A version of AlphaNumericCompare() for comparing UTF8 encoded data
+// Working with bytes is 10 times faster than converting to wstring
```

Comment:
Shall this comment be in Doxygen style? So that the difference between this and AlphaNumericCompare can be seen in the code document.

Reply:
No, Doxygen comments (the minority of things that have then) generally go in the header file. Such comments on the older code that is not part of Skin or plugin API are very sparse, and as a result IMO the code doc generated are pretty useless. Most that exist are trivail too, doing little more than repeat the function name. Anyone working seriously with core will need to read through the code, and this comment is exactly where it is likely to be most useful.

Kodi could benefit greatly from overall design documentation, but as yet no one has volunteered to create it.